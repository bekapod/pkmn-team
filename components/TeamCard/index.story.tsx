// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from "@storybook/react";
import React from "react";
import TeamCard from ".";
import { Team } from "../../types";

const team: Team = {
  insertedAt: "2018-06-08T21:15:14.723Z",
  id: "1",
  members: [
    {
      id: "1",
      order: 1,
      pokemon: {
        id: "25",
        name: "Pikachu",
        slug: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "Electric", slug: "electric" }]
      }
    },
    {
      id: "2",
      order: 2,
      pokemon: {
        id: "25",
        name: "Pikachu",
        slug: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "Electric", slug: "electric" }]
      }
    },
    {
      id: "3",
      order: 3,
      pokemon: {
        id: "25",
        name: "Pikachu",
        slug: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "Electric", slug: "electric" }]
      }
    }
  ],
  name: "Pikachu Team"
};

storiesOf("TeamCard", module).add("default", () => <TeamCard team={team} />);
