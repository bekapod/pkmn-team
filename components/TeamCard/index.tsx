import dateFormat from "dateformat";
import { compose, flatMap, get, isNil, map, reject } from "lodash/fp";
import Link from "next/link";
import React from "react";
import * as variables from "../../helpers/variables";
import { Pokemon, Team } from "../../types";
import { CardContent, CardHeader, CardLink, CardWrapper } from "../Card";
import CardHeading from "../CardHeading";
import CardMeta from "../CardMeta";
import PokemonLine from "../PokemonLine";

const getAllTypes = flatMap(get("types"));

interface Props {
  team: Team;
}

const TeamCard = ({
  team: { id, name, members, createdAt }
}: Props): JSX.Element => {
  const pokemon: Pokemon[] = compose([reject(isNil), map(get("pokemon"))])(
    members
  );

  return (
    <Link href={`/team/edit/${id}/`} passHref>
      <CardLink data-testid={`team-link-${id}`}>
        <CardWrapper>
          <CardHeader types={getAllTypes(pokemon)}>
            <CardHeading>{name}</CardHeading>
          </CardHeader>

          <CardContent>
            <CardMeta
              id={id}
              items={[
                { label: "Pkmn", value: members.length },
                { label: "Created", value: dateFormat(createdAt, "d/m/yy") }
              ]}
            />

            {members.map(
              ({ id: memberId, pokemon: memberPkmn }): JSX.Element => (
                <PokemonLine
                  key={`Team Member: ${memberId}`}
                  pokemon={memberPkmn}
                  outdent={variables.spacing.sm}
                />
              )
            )}
          </CardContent>
        </CardWrapper>
      </CardLink>
    </Link>
  );
};

export default TeamCard;
