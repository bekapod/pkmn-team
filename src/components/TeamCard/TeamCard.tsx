import dateFormat from 'dateformat';
import compose from 'lodash/fp/compose';
import flatMap from 'lodash/fp/flatMap';
import get from 'lodash/fp/get';
import isNil from 'lodash/fp/isNil';
import map from 'lodash/fp/map';
import reject from 'lodash/fp/reject';
import Link from 'next/link';
import { FunctionComponent } from 'react';
import {
  PokemonFragmentFragment,
  Teams,
  Team_Member
} from '~/generated/graphql';
import {
  CardContent,
  CardHeader,
  CardHeading,
  CardLink,
  CardWrapper
} from '../Card';
import { CardMeta } from '../CardMeta';
import { PokemonLine } from '../PokemonLine';

const getAllTypes = compose(flatMap(get('type')), flatMap(get('types')));

type TeamMember = Pick<Team_Member, 'id' | 'order'> & {
  pokemon: PokemonFragmentFragment;
};
export type TeamCardProps = Pick<Teams, 'id' | 'name' | 'created_at'> & {
  team_members: TeamMember[];
};

export const TeamCard: FunctionComponent<TeamCardProps> = ({
  id,
  name,
  team_members,
  created_at
}) => {
  const pokemon: PokemonFragmentFragment[] = compose([
    reject(isNil),
    map(get('pokemon'))
  ])(team_members);

  return (
    <Link href={`/team/${id}/`} passHref>
      <CardLink data-testid={`team-link-${id}`}>
        <CardWrapper>
          <CardHeader types={getAllTypes(pokemon)}>
            <CardHeading>{name}</CardHeading>
          </CardHeader>

          <CardContent>
            <CardMeta
              id={id}
              items={[
                { label: 'Pkmn', value: team_members.length },
                { label: 'Created', value: dateFormat(created_at, 'd/m/yy') }
              ]}
            />

            {team_members.map(
              ({ id: memberId, pokemon: memberPkmn }): JSX.Element => (
                <PokemonLine
                  key={`Team Member: ${memberId}`}
                  pokemon={memberPkmn}
                  outdent="var(--spacing-3)"
                />
              )
            )}
          </CardContent>
        </CardWrapper>
      </CardLink>
    </Link>
  );
};
