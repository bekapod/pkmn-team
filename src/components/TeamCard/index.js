// @flow
import React from "react";
import { compose, map, flatten, prop, propOr, length } from "ramda";
import dateFormat from "dateformat";
import { CardWrapper, CardHeader, CardContent } from "../Card";
import CardHeading from "../CardHeading";
import CardMeta from "../CardMeta";
import PokemonLine from "../PokemonLine";
import * as variables from "../../helpers/variables";
import type { Team } from "../../types";

const getAllTypes = compose(
  flatten,
  map(propOr([], "types"))
);

type Props = {
  team: Team
};

const TeamCard = ({ team: { id, name, members, createdAt } }: Props) => {
  const pokemon = members.map(member => prop("pokemon", member));

  return (
    <CardWrapper>
      <CardHeader types={getAllTypes(pokemon)}>
        <CardHeading>{name}</CardHeading>
      </CardHeader>

      <CardContent>
        <CardMeta
          id={id}
          items={[
            { label: "Pkmn", value: length(members) },
            { label: "Created", value: dateFormat(createdAt, "d/m/yy") }
          ]}
        />

        {members.map(({ id: memberId, pokemon: memberPkmn }) => (
          <PokemonLine
            key={`Team Member: ${memberId}`}
            pokemon={memberPkmn}
            outdent={variables.spacing.sm}
          />
        ))}
      </CardContent>
    </CardWrapper>
  );
};

export default TeamCard;
