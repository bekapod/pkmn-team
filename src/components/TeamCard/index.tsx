import dateFormat from "dateformat";
import { compose, flatMap, get, isNil, map, reject } from "lodash/fp";
import React from "react";
import * as variables from "../../helpers/variables";
import { IPokemon, ITeam, Type } from "../../types";
import { CardContent, CardHeader, CardWrapper } from "../Card";
import CardHeading from "../CardHeading";
import CardMeta from "../CardMeta";
import PokemonLine from "../PokemonLine";

const getAllTypes = flatMap(get("types"));

interface IProps {
  team: ITeam;
}

const TeamCard = ({ team: { id, name, members, createdAt } }: IProps) => {
  const pokemon: IPokemon[] = compose([reject(isNil), map(get("pokemon"))])(
    members
  );

  return (
    <CardWrapper>
      <CardHeader types={getAllTypes(pokemon)}>
        <CardHeading>{name}</CardHeading>
      </CardHeader>

      <CardContent>
        <CardMeta
          id={id}
          items={[
            { label: "Pkmn", value: members.length },
            { label: "Created", value: dateFormat(createdAt, "d/m/yy") }
          ]}
        />

        {members.map(({ id: memberId, pokemon: memberPkmn }) => (
          <PokemonLine
            key={`Team Member: ${memberId}`}
            pokemon={memberPkmn}
            outdent={variables.spacing.sm}
          />
        ))}
      </CardContent>
    </CardWrapper>
  );
};

export default TeamCard;
