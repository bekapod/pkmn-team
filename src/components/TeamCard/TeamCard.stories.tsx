import { Meta } from '@storybook/react/types-6-0';
import { Pokemon } from '~/generated/graphql';
import { charmander, haunter, pikachu } from '~/mocks/Pokemon';
import { TeamCard, TeamCardProps } from './TeamCard';

const pokemon: Pokemon[] = [charmander, pikachu, haunter];

export default {
  title: 'Components/TeamCard',
  component: TeamCard,
  args: {
    id: '123',
    created_at: '2020-12-12T22:50:59.766899+00:00',
    name: 'My super team!',
    team_members: [
      {
        id: '1',
        order: 1,
        pokemon: pokemon[0]
      },
      {
        id: '2',
        order: 2,
        pokemon: pokemon[1]
      },
      {
        id: '3',
        order: 3,
        pokemon: pokemon[2]
      }
    ]
  }
} as Meta<TeamCardProps>;

export const Standard = (args: TeamCardProps): JSX.Element => (
  <TeamCard {...args} />
);
