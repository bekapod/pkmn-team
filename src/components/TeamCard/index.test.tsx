import React from "react";
// tslint:disable-next-line:no-implicit-dependencies
import renderer from "react-test-renderer";
import TeamCard from ".";

describe("<TeamCard />", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(
      <TeamCard
        team={{
          createdAt: "2018-06-08T21:15:14.723Z",
          id: "1",
          members: [
            {
              id: "1",
              pokemon: {
                id: "25",
                name: "pikachu",
                pokedexId: 25,
                sprite: "25.png",
                types: ["ELECTRIC"]
              }
            },
            {
              id: "2",
              pokemon: {
                id: "25",
                name: "pikachu",
                pokedexId: 25,
                sprite: "25.png",
                types: ["ELECTRIC"]
              }
            },
            {
              id: "3",
              pokemon: {
                id: "25",
                name: "pikachu",
                pokedexId: 25,
                sprite: "25.png",
                types: ["ELECTRIC"]
              }
            }
          ],
          name: "Pikachu Team"
        }}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
