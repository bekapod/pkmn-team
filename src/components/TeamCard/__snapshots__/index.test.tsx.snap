// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TeamCard /> renders without crashing 1`] = `
<article
  className="Card__CardWrapper-sc-3o7tcb-0 bTjMad"
>
  <header
    className="Card__CardHeader-sc-3o7tcb-1 kYOeEy"
  >
    <h2
      className="CardHeading-mmeo34-0 ebVIiR"
    >
      Pikachu Team
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-3o7tcb-2 fDtTnd"
  >
    <div
      className="CardMeta__CardMetaWrapper-yqfxft-0 kTYlws"
    >
      <div
        className="CardMeta__CardMetaItem-yqfxft-1 hPoDTe"
      >
        <span
          className="CardMeta__CardMetaLabel-yqfxft-2 hiNCFT"
        >
          Pkmn
        </span>
        <span
          className="CardMeta__CardMetaValue-yqfxft-3 jqYdyo"
        >
          3
        </span>
      </div>
      <div
        className="CardMeta__CardMetaItem-yqfxft-1 hPoDTe"
      >
        <span
          className="CardMeta__CardMetaLabel-yqfxft-2 hiNCFT"
        >
          Created
        </span>
        <span
          className="CardMeta__CardMetaValue-yqfxft-3 jqYdyo"
        >
          8/6/18
        </span>
      </div>
    </div>
    <div
      className="PokemonLine__Row-sc-1qiwepa-0 dAGLvI"
    >
      <img
        alt="pikachu sprite"
        className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
        src="/sprites/25.png"
      />
      <div>
        <div
          className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
        >
          #25 Pikachu
        </div>
        <ul
          className="InlineList-sc-1dn69qz-0 YzWxJ"
        >
          <li>
            <span
              className="TypeTag-sc-1dttu8y-0 byzsso"
              type="ELECTRIC"
            >
              ELECTRIC
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="PokemonLine__Row-sc-1qiwepa-0 dAGLvI"
    >
      <img
        alt="pikachu sprite"
        className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
        src="/sprites/25.png"
      />
      <div>
        <div
          className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
        >
          #25 Pikachu
        </div>
        <ul
          className="InlineList-sc-1dn69qz-0 YzWxJ"
        >
          <li>
            <span
              className="TypeTag-sc-1dttu8y-0 byzsso"
              type="ELECTRIC"
            >
              ELECTRIC
            </span>
          </li>
        </ul>
      </div>
    </div>
    <div
      className="PokemonLine__Row-sc-1qiwepa-0 dAGLvI"
    >
      <img
        alt="pikachu sprite"
        className="PokemonLine__RowImage-sc-1qiwepa-1 hUcFnr"
        src="/sprites/25.png"
      />
      <div>
        <div
          className="PokemonLine__RowTitle-sc-1qiwepa-2 edSfRl"
        >
          #25 Pikachu
        </div>
        <ul
          className="InlineList-sc-1dn69qz-0 YzWxJ"
        >
          <li>
            <span
              className="TypeTag-sc-1dttu8y-0 byzsso"
              type="ELECTRIC"
            >
              ELECTRIC
            </span>
          </li>
        </ul>
      </div>
    </div>
  </div>
</article>
`;
